### PR TITLE
#4700 Seed Input Fields for RandomInt and RandomFloat

### LINKED ISSUE #4700: [enhancement]: Random nodes should have seed fields
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Contact Details

_No response_

### What should this feature add?

Random nodes (like random integer, random float) should have a seed field to seed the RNG.

Without this, workflows/graphs are not deterministic.

### Alternatives

_No response_

### Additional Content

_No response_

### PR DESCRIPTION
**Pull request by glibesyck**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Description

RandomInt and RandomFloat Nodes now have seed input field for reproducible workflows.

## Related Tickets & Documents

closes #4700 
relates to #4581

### CODE REVIEW

**Comment by Millu** [comment]
Thanks for adding the seeds! 

As I was testing this, realized with a seed field in the RNG nodes, there's no easy way to have random generations via workflows now, unless a user manually changes the seed after Invoke-ing each time. 

Some options after discussing with @psychedelicious 
* Adopt a a seed modification system similar to Comfy
* "Random Seed" node - would have CPU / GPU option (it's seeds all the way down)
*  Remove seed option from RNG nodes and embed seed in image metadata graph for reproducibility 
* Move seed generation to the backend to function the same across Linear UI & workflows

Would love to get your thoughts @glibesyck!

**Comment by glibesyck** [comment]
Thank you for the comments and discussion @Millu!
Great, I will give it a look at how Comfy reaches the desired functionality and return back;)

**Comment by glibesyck** [comment]
After some research time, it seems to me that keeping seed value in image metadata is the most natural way for user to look for these values and fix them if needed. I've found the similar issue #4495 which addresses this question from more generalized perspective, including this one as well. I'm not sure whether I'm able to assist in development for this one, still it seems to me that it is indeed the great way to account for reproducibility.

File: invokeai/app/invocations/math.py [hunk 1]
```diff
@@ -63,13 +65,15 @@ def invoke(self, context: InvocationContext) -> IntegerOutput:
     use_cache=False,
 )
 class RandomIntInvocation(BaseInvocation):
-    """Outputs a single random integer."""
+    """Outputs a single random integer (with possibility to set seed)."""
 
     low: int = InputField(default=0, description=FieldDescriptions.inclusive_low)
     high: int = InputField(default=np.iinfo(np.int32).max, description=FieldDescriptions.exclusive_high)
+    seed: int = InputField(default=None, ge=0, le=SEED_MAX, description=FieldDescriptions.seed_random_generation)
```
Comment (psychedelicious) [inline invokeai/app/invocations/math.py:72]: This should have a default value of a number (probably zero)

File: invokeai/app/invocations/math.py [hunk 2]
```diff
@@ -79,10 +83,12 @@ class RandomFloatInvocation(BaseInvocation):
     low: float = InputField(default=0.0, description=FieldDescriptions.inclusive_low)
     high: float = InputField(default=1.0, description=FieldDescriptions.exclusive_high)
     decimals: int = InputField(default=2, description=FieldDescriptions.decimal_places)
+    seed: int = InputField(default=None, ge=0, le=SEED_MAX, description=FieldDescriptions.seed_random_generation)
```
Comment (psychedelicious) [inline invokeai/app/invocations/math.py:86]: Default value should be an int

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Thanks for working on this @glibesyck !

The actual code changes here are fine I think, but this would be a breaking change as implemented. We'll need the seeded RNG to be opt-in, so the previous behaviour is not altered. That could be behind a boolean field.

Note that fields with "direct" input - like a number input on the UI - will never be `None`. Even if you set the seed field to `Optional[int] = InputField(default=None...`, the node would *always* get an integer. So whether or not the RNG is seeded cannot depend on the seed value itself, it *must* be a separate boolean field.

**Comment by psychedelicious** [comment]
This PR does raise a deeper issue for which the solution is not clear, and that's what @Millu was asking about. It's not so much a question of how the seed value should be stored, but how we introduce randomness to the graph.

I've added a comment with more details in #4700 in case you have any thoughts.

**Comment by Millu** [comment]
@glibesyck thank you for your effort on this! Going to close this for now as it appears this will be address in an alternative method 

We appreciate your contributions and would love to continue to see them, especially if there are any other items of interest here: https://github.com/orgs/invoke-ai/projects/7/views/10

**Pull request closed without merging**